### PR TITLE
Gps parsing

### PR DESCRIPTION
**Pull request by regeliv**
- Implement basic GPS data parsing using the TinyGPS++ library.
- Some miscellaneous QoL improvements

### CODE REVIEW

File: projects/T-beam/include/gps.h [hunk 1]
```diff
@@ -4,8 +4,21 @@
 #define GPS_TX_PIN 12
```
Comment (Serferrr) [inline projects/T-beam/include/gps.h:4]: don't use #define

File: projects/T-beam/src/gps.cpp [hunk 2]
```diff
@@ -1,26 +1,38 @@
...
+        gps_data.lng = tiny_gps.location.lng();
+        gps_data.time.hours = tiny_gps.time.hour();
+        gps_data.time.minutes = tiny_gps.time.minute();
+        gps_data.time.seconds = tiny_gps.time.second();
+
+        data_available = true;
+      }
     }
+
+    vTaskDelay(pdMS_TO_TICKS(100));
```
Comment (Serferrr) [inline projects/T-beam/src/gps.cpp:35]: are we 100% sure this value will always be 100? Is there any benefit to making it a variable?

File: projects/T-beam/src/main.cpp [hunk 3]
```diff
@@ -3,12 +3,27 @@
...
+  for (;;) {
+    if (!gps::data_is_available()) {
+      Serial.println("GPS data is unavailable.");
+    } else {
+      gps::Data gps_data = gps::get_gps_data();
+      Serial.printf("Lat: %.6f Long: %.6f Time: %02d:%02d:%02d\n", gps_data.lat,
+                    gps_data.lng, gps_data.time.hours, gps_data.time.minutes,
+                    gps_data.time.seconds);
+    }
+    vTaskDelay(pdMS_TO_TICKS(1000));
```
Comment (Serferrr) [inline projects/T-beam/src/main.cpp:16]: same as above, will it always be 1000?

**Review by Serferrr — CHANGES_REQUESTED** [review]

File: projects/T-beam/src/main.cpp [hunk 4]
```diff
@@ -3,12 +3,27 @@
 #include "gps.h"
 #include "led.h"
 
+void log(void *pvParameters) {
```
Comment (Serferrr) [inline projects/T-beam/src/main.cpp:6]: this function shouldn't be here, move it to an appropriate place

File: projects/T-beam/include/gps.h (hunk 1 above)
Comment (regeliv) [reply projects/T-beam/include/gps.h:4]: IMO the usage of defines should be addressed in a seperate pull request, given that there are a few of them in the codebase

File: projects/T-beam/src/gps.cpp (hunk 2 above)
Comment (regeliv) [reply projects/T-beam/src/gps.cpp:35]: At the moment it's essentially a placeholder, no reason to refactor into a variable. Possibly, we could use interrupts to read the GPS serial, but right now the focus was on getting basic functionality

File: projects/T-beam/src/main.cpp (hunk 3 above)
Comment (regeliv) [reply projects/T-beam/src/main.cpp:16]: It's a placeholder, similar to above.

File: projects/T-beam/src/main.cpp (hunk 4 above)
Comment (regeliv) [reply projects/T-beam/src/main.cpp:6]: This function will eventually be used to log data to a disk from all components. Given that the main file will naturally have to aggregate these, the placement of this function is appropriate.

File: projects/T-beam/include/gps.h (hunk 1 above)
Comment (Serferrr) [reply projects/T-beam/include/gps.h:4]: Don't defer work that can be done now for later, it's true that those other defines need to be addressed but this one can, and should be, changed now

File: projects/T-beam/src/main.cpp (hunk 4 above)
Comment (Serferrr) [reply projects/T-beam/src/main.cpp:6]: If so please add a comment saying why this function is where it is and how will it change in the future to be differently than it is now

**Comment by Bartosz-Slowik** [comment]
Usage of define is not wrong, 
There is not really a difference using a cost here except for a known type

We used defines in the rest of codebase so I would prefer it to be define for now and refactored later.

It's not being lazy it's having code integrity,  I prefer having consistent syntax over some marginal personal preference.

I also prefer defines at it restricts some bad ideas like taking an address to a static and it's immediately known when reading code

We can use either tbh.

**Pull request merged**